internal/controlplane/handlers_repositories.go

```diff
@@ -308,29 +309,61 @@ func (s *Server) ListRemoteRepositoriesFromProvider(
 	entityCtx := engine.EntityFromContext(ctx)
 	projectID := entityCtx.Project.ID
 
-	provider, err := getProviderFromRequestOrDefault(ctx, s.store, in, projectID)
+	provs, err := listProvidersOrInferDefault(ctx, s.store, in, projectID)
 	if err != nil {
 		return nil, providerError(err)
 	}
 
-	zerolog.Ctx(ctx).Debug().
-		Str("provider", provider.Name).
-		Str("projectID", projectID.String()).
-		Msg("listing repositories")
-
-	pbOpts := []providers.ProviderBuilderOption{
-		providers.WithProviderMetrics(s.provMt),
-		providers.WithRestClientCache(s.restClientCache),
-	}
-	p, err := providers.GetProviderBuilder(ctx, provider, s.store, s.cryptoEngine, &s.cfg.Provider, pbOpts...)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "cannot get provider builder: %v", err)
+	out := &pb.ListRemoteRepositoriesFromProviderResponse{
+		Results: []*pb.UpstreamRepositoryRef{},
 	}
 
-	if !p.Implements(db.ProviderTypeRepoLister) {
-		return nil, util.UserVisibleError(codes.Unimplemented, "provider does not implement repository listing")
+	failIfListerNotImplemented := len(provs) == 1
+
+	for _, provider := range provs {
+		zerolog.Ctx(ctx).Debug().
+			Str("provider", provider.Name).
+			Str("projectID", projectID.String()).
+			Msg("listing repositories")
+
+		pbOpts := []providers.ProviderBuilderOption{
+			providers.WithProviderMetrics(s.provMt),
+			providers.WithRestClientCache(s.restClientCache),
+		}
+		p, err := providers.GetProviderBuilder(ctx, provider, s.store, s.cryptoEngine, &s.cfg.Provider, pbOpts...)
```

Comment:
Not for this PR, but it feels like most (5 of 7) of the arguments to `GetProviderBuilder` are actually associated with the server, so we should think about having a "ProviderBuilderFactory" (or some better name, like `ProviderClass`) that takes the `store`, `cryptoEngine`, etc arguments, and is stored as a field on the `server`.

WDYT?

Reply:
Good idea, let's do that in another PR.